### PR TITLE
Splitting out general purpose helpers for branches and loops that yield values

### PR DESCRIPTION
**Pull request by bettinaheim**
Fixes https://github.com/microsoft/qsharp-compiler/issues/974.
As part of fixing https://github.com/microsoft/qsharp-compiler/issues/974, I decided to split out the necessary logic into some general purpose methods. 
I also decided to get rid of the final comma after all for array printing (as previously suggested by @kuzminrobin), though it does come at the cost of having to evaluate a branching in each loop iteration (I am still not sure whether that is preferable or not).

### CODE REVIEW

File: src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs [hunk 1]
```diff
@@ -544,7 +548,7 @@ Value TupleToString(TupleValue tuple)
                         }
                         else
                         {
-                            comma ??= CreateConstantString(",");
+                            comma ??= CreateConstantString(", ");
                             str = DoAppend(str, comma, unreferenceNext: false);
                         }
                     }
```
Comment (kuzminrobin) [inline src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs:552]: FYI: Another way to construct a comma-separated list in parentheses is about like this:
```C++
// Output an opening parenthesis, or `$"{tuple.TypeName}("`.

for (var idx = 0; idx < tupleElements.Length; ++idx)
{
    if(idx != 0)
    {
        // Output comma (`, `).
    }
    // Output `element[i]`.
}

// Output the closing parenthesis `)`.
```

File: src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs [hunk 2]
```diff
@@ -1199,7 +1157,10 @@ public override ResolvedExpressionKind OnConditionalExpression(TypedExpression c
             }
             else
             {
-                var evaluated = this.ConditionalEvaluation(cond.Value, onCondTrue: ifTrueEx, onCondFalse: ifFalseEx);
+                var evaluated = this.SharedState.ConditionalEvaluation(
+                    cond.Value,
+                    onCondTrue: () => this.SharedState.EvaluateSubexpression(ifTrueEx),
+                    onCondFalse: () => this.SharedState.EvaluateSubexpression(ifFalseEx));
```
Comment (kuzminrobin) [inline src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs:1161]: nit: I prefer aligning the parts of the similar strings such that it is immediately visible what's the difference between the two strings, like:
```c#
                    onCondTrue : () => this.SharedState.EvaluateSubexpression(ifTrueEx ),
                    onCondFalse: () => this.SharedState.EvaluateSubexpression(ifFalseEx));
```

**Review by kuzminrobin — APPROVED** [review]

File: src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs (hunk 2 above)
Comment (bettinaheim) [reply src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs:1161]: Personally I also like that very much. However, it is not compatible with the C# formatting rules that are enforced (or the F# ones for that matter...).

File: src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs (hunk 1 above)
Comment (bettinaheim) [reply src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs:551]: Sure, either way is fine for me, so I changed it.

**Pull request merged**